src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -546,33 +546,73 @@ private void doScheduledSearchForRound (RaptorState outputState) {
+                        // be preferred. This will not affect total travel time, but it will affect the breakdown of
+                        // walk vs. wait. So if needed as an optimization, these blocks could be disabled unless
+                        // request.computePaths = true;
+                        boolean boardHereInstead = false;
+                        if (inputState.previous == null) {
+                            // In the first transit round, arriving at this stop before arriving at the upstream stop
+                            // implies shorter access time. And arriving earlier or at the same time implies more
+                            // buffer time to board the trip, so boarding at this stop is preferable to boarding at
+                            // the upstream stop when the non-strict inequality is satisfied.
+                            if (inputState.bestTimes[stop] <= inputState.bestTimes[boardStop]) {
+                                boardHereInstead = true;
+                            }
+                        } else {
+                            // In subsequent transit rounds, we want to compare the length of the transfer legs.
+                            // Boarding a trip at this stop is preferable if it requires a shorter transfer than
+                            // boarding the same trip at the upstream stop.
+                            int upstreamTransferFromStop = inputState.transferStop[boardStop];
+                            if (upstreamTransferFromStop != -1) {
+                                int upstreamTransferTime =
+                                        inputState.bestTimes[boardStop] - inputState.previous.bestTimes[upstreamTransferFromStop];
+                                int transferFromStop = inputState.transferStop[stop];
+                                int transferTime =
+                                        inputState.bestTimes[stop] - inputState.previous.bestTimes[transferFromStop];
+                                if (transferTime < upstreamTransferTime) {
+                                    boardHereInstead = true;
+                                }
+                            }
+                        }
+
+                        if (boardHereInstead) {
+                            boardTime = onTripSchedule.departures[stopPositionInPattern];
+                            waitTime = onTripSchedule.departures[stopPositionInPattern] - inputState.bestTimes[stop];
+                            boardStop = stop;
+                        }
+
+                        // Second, it might even be possible to board an earlier trip at the downstream stop. This
+                        // must be checked to enssure correctness of total travel time results.
+                        // TODO handle trips on the same pattern that overtake each other?
+                        int earlierTripIdx = onTrip;
+                        while (--earlierTripIdx >= 0) {
```

Comment:
It strikes me looking at this that once we're on a trip, we look at at least one earlier trip on every single downstream stop. This checking of earlier trips is in fact only necessary if the previous-round arrival time at that stop is earlier than (less than, not equal to) the departure time on the trip we're currently riding. I guess checking that additional condition is an optimization and may not actually save much computation over stepping back one trip at every stop and checking whether that previous trip is boardable.

Additionally, if we board an earlier trip, then all the decision logic about whether to boardHereInstead was in vain. So ideally that boardHereInstead logic would only run once it's clear we won't board an earlier trip. But again this is an optimization and may have negligible impact. I think it might just better reflect the structure of the decision tree.